### ISSUE TITLE
SYSTEMD_WANTS does not preserve backslashes

### ISSUE THREAD

**Issue by JakubVanek** [systemd/systemd#16735]
**systemd version the issue has been seen with**
241-7~deb10u3 (but this is likely applicable to `master` as well)

**Used distribution**
ev3dev-buster (~= Debian buster)

**Expected behaviour you didn't see**
Backslashes are preserved in unit names coming from udev `SYSTEMD_WANTS`. The following udev rule should start a unit named `ev3-uart@ev3\x2dports:in4.service` if `LEGO_ADDRESS` is `ev3-ports:in4`:
```
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    PROGRAM="/bin/systemd-escape $env{LEGO_ADDRESS}" \
    ENV{SYSTEMD_WANTS}="ev3-uart@$result.service", \
    ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
```

**Unexpected behaviour you saw**
Backslashes are stripped, therefore breaking manual escaping. Unit named `ev3-uart@ev3x2dports:in4.service` is started instead of the one above. The template unit is then unable to unescape the `ev3x2dports:in4` instance name. However, the unit requires both the escaped and unescaped names:
```
# This file is part of the ev3-systemd package

[Unit]
Description=LEGO MINDSTORMS EV3 UART Sensors
# This is the alias created by the ev3dev-uart udev rule.
BindsTo=sys-subsystem-lego-devices-ev3_uart_host_%i.device
Wants=systemd-udev-settle.service
After=systemd-udev-settle.service

[Service]
Type=forking
ExecStart=/usr/sbin/ldattach 29 /dev/tty_%I
```

I think the problem is in https://github.com/systemd/systemd/blob/61c8b73a35fd71cb7cbb41b78d93a124bf347e8e/src/core/device.c#L379, where the ~quote~ backslash may be stripped as part of unquoting.

Hardcoding `SYSTEMD_WANTS` to contain a second backslash does work around the problem, but this is more difficult to do automatically inside the udev rule.

**Steps to reproduce the problem**
 * Find a device that can be hot-swapped
 * Create a template systemd unit that depends on both `%i` and `%I` and accepts a non-sysfs escaped string as an instance name
 * Add a udev rule for the device with a `SYSTEMD_WANTS` property that either uses `systemd-escape` result or hardcodes the instance name with only one slash
 * Observe that the unit started on device hotplug is not able to reverse the escaping done by `systemd-escape`

**Related bugs**
OP in https://github.com/systemd/systemd/issues/7109 observed something similar. Using empty instance name as introduced in https://github.com/systemd/systemd/commit/dcebc9bae4dcc3e844f01558c6127fc0d8745c8e would be another way to work around this, albeit not an ideal one.

**Comment by JakubVanek** [comment]
Udev itself doesn't seem to be the problem, `/run/udev/data/+lego\:ev3-ports\:in4\:ev3-uart-host` does contain the backslash:
```
I:4505815485
E:SYSTEMD_WANTS=ev3-uart@ev3\x2dports:in4.service
E:SYSTEMD_ALIAS=/sys/subsystem/lego/devices/ev3_uart_host_ev3-ports:in4
G:systemd
```

**Comment by yuwata** [comment]
I hope this is fixed by #16737.

**Comment by TriMoon** [comment]
I think the real problem is in the udev rule.
> ```
> SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
>     PROGRAM="/bin/systemd-escape $env{LEGO_ADDRESS}" \
>     ENV{SYSTEMD_WANTS}="ev3-uart@$result.service", \
>     ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
> ```
Here `PROGRAM` is set to the escaped value of `LEGO_ADDRESS` but the unescaped value is used in setting `SYSTEMD_ALIAS`....

**Comment by JakubVanek** [comment]
I first thought this as well, but then I couldn't explain why `/run/udev/data/+lego:ev3-ports:in4:ev3-uart-host` contained `SYSTEMD_ALIAS` with the backslash: `E:SYSTEMD_WANTS=ev3-uart@ev3\x2dports:in4.service`. However, there is indeed some processing happening in udev as well. I couldn't add the second backslash using sed or an external script, the preceding backslashes would always get replaced by `_` (like `ev3-uart@ev3_\x2dports:in4.service`).

I'm currently trying to build a systemd debian package with the fix to test, but until recently I was stuck with `version.h` not being generated inside a `pbuilder` session. Manually creating it during the build seemed to fix that, so hopefully it will finish successfully now.

**Comment by TriMoon** [comment]
> the preceding backslashes would always get replaced by `_` (like `ev3-uart@ev3_\x2dports:in4.service`).

That would indicate that there is a space slipped in before `-ports:in4.service` in which case it is indeed translated to a space plus escaped `-` :wink:

**Comment by JakubVanek** [comment]
> > the preceding backslashes would always get replaced by `_` (like `ev3-uart@ev3_\x2dports:in4.service`).
> 
> That would indicate that there is a space slipped in before `-ports:in4.service` in which case it is indeed translated to an `_` (space) plus escaped `-` (\x2d) :wink:

I haven't thought about that, I will try to do some experiments.

> https://github.com/systemd/systemd/blob/61c8b73a35fd71cb7cbb41b78d93a124bf347e8e/src/core/device.c#L379
> 
> Is just to translate escaped names back to their original form to keep the unit files clean ASCII if im not mistaken.
Because one could use filenames in unicode depending on underlying FS....

I think that mangled->unicode conversion is done in `unit_name_path_unescape()`, as this is what the command `systemd-escape --path --unescape` calls. `extract_first_word` in this case should only do unquoting, as I've seen nested quotes like `SYSTEMD_WANTS+="'%c'"` in some guides on the internet.

**Comment by TriMoon** [comment]
You could be right, just trying to shed another light from my eyes to the problem at hand :smiley_cat: 
I just dug through the `systemd-escape` man page and that shows it will translate spaces to `\x20` so it must be injected elsewhere...

**Comment by JakubVanek** [comment]
The `_` characters do not seem come from spaces, at least not initially (but it may indeed by that udev translates other characters to spaces internally).
To prevent escaping issues inside udev rules, I've created an external script to supply the name to udev (`/bin/print-udev-name.sh`):
```sh
#!/bin/sh
echo 'ev3-uart@ev3\\\\x2dports:in4.service'
exit 0
```
```sh
$ /bin/print-udev-name.sh
ev3-uart@ev3\\x2dports:in4.service
```
Then I changed the udev rule so that it used the script result directly.
```udev
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    PROGRAM="/bin/print-udev-name.sh" \
    ENV{SYSTEMD_WANTS}="$result", \
    ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
```
The result is that the underscore is still added to the udev data:
```
$ cat /run/udev/data/+lego\:ev3-ports\:in4\:ev3-uart-host 
I:1191259570
E:SYSTEMD_WANTS=ev3-uart@ev3_\x2dports:in4.service
E:SYSTEMD_ALIAS=/sys/subsystem/lego/devices/ev3_uart_host_ev3-ports:in4
G:systemd
```
But interestingly, when writing the two backslashes directly into `SYSTEMD_WANTS`, it works correctly:
```udev
SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
    ENV{SYSTEMD_WANTS}="ev3-uart@ev3\\x2dports:in4.service", \
    ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
```
```
$ cat /run/udev/data/+lego\:ev3-ports\:in4\:ev3-uart-host 
I:1520068121
E:SYSTEMD_WANTS=ev3-uart@ev3\\x2dports:in4.service
E:SYSTEMD_ALIAS=/sys/subsystem/lego/devices/ev3_uart_host_ev3-ports:in4
G:systemd
```

**Comment by TriMoon** [comment]
Well then there we go, issue at topic is not the cause, next step = udev issue tracker right? :wink:

**Comment by JakubVanek** [comment]
> Well then there we go, issue at topic is not the cause, next step = udev issue tracker right? wink

Why? Systemd should work with `E:SYSTEMD_WANTS=ev3-uart@ev3\x2dports:in4.service` (single backslash) and it currently doesn't (which is what this bugreport is about). This is what udev is able to generate. The fact that I cannot workaround the systemd bug through backslash doubling in `PROGRAM` in udev rule is another thing (that may well work as specified, just I haven't tried finding the specification).

**Comment by TriMoon** [comment]
Then i would suggest you to change the title into "**systemd_wants should accept unescaped device names**", because that is what you actually want.
And i think that behavior of using escaped device names is used in other keywords as well, so im not sure if they want to change that API...

A single backslash is interpreted as an escape character when used **inside** double quotes by all shells (which is used to parse the udev rules IIRC), while a double backslash translates to a single backslash :wink: 
That's why i concluded that it's not actually an issue with systemd...

**Comment by JakubVanek** [comment]
> Then i would suggest you to change the title into "**systemd_wants should accept unescaped device names**", because that is what you actually want.
> And i think that behavior of using escaped device names is used in other keywords as well, so im not sure if they want to change that API...
> 
> A single backslash is interpreted as an escape character when used **inside double quotes** by all shells (which is used to parse the udev rules if im not mistaken), while a double backslash translates to a single backslash :wink:
> eg.:
> 
>     * `ENV{SYSTEMD_WANTS}="ev3-uart@ev3\x2dports:in4.service"` gets translated into:
>       `export SYSTEMD_WANTS='ev3-uart@ev3-ports:in4.service'` (Note the dash)
> 
>     * `ENV{SYSTEMD_WANTS}="ev3-uart@ev3\\x2dports:in4.service"` gets translated into:
>       `export SYSTEMD_WANTS='ev3-uart@ev3\x2dports:in4.service'`
> 
>     * And `ev3-uart@ev3\x2dports:in4.service` is the correct form expected by systemd's `Wants=`...
> 

From this POV the current systemd behaviour does make sense and the fix in https://github.com/systemd/systemd/pull/16737 introduces a regression. I haven't realized that `EXTRACT_RETAIN_ESCAPE` breaks `"\""` and this unescape was definitely meant to work (with the fix, it would get translated to `\"`). I guess the same way the `\x` got translated to simple unescaped `x` without further interpretation (~ C unescaping). Indeed, "systemd_wants should accept unescaped device names" would be a better fit.

> Maybe you need to change the instance-template unit `ev3-uart@.service` to accept unescaped instance names?
> eg. expect `ev3-ports:in4` instead of `ev3\x2dports:in4` as an instance name...

I don't see a way of doing it without further changes. The systemd-encoded device path is required for its `BindsTo` parameter: https://github.com/ev3dev/ev3-systemd/blob/ev3dev-stretch/systemd/ev3-uart%40.service The instance name is then un-encoded and passed to ldattach to start the driver.

There is a way to work this around through empty instance name, but I was hoping that could be avoided. If a template like `ev3-uart@.service` is detected in the udev `SYSTEMD_WANTS` field, systemd automatically replces the empty instance name with systemd-encoded sysfs path to the device. https://github.com/ev3dev/ev3dev/issues/1433#issuecomment-674375924

**Comment by JakubVanek** [comment]
I will check if I can pass the double backslash from udev by replacing `\` with `\x5C\x5C`, but I'm afraid that systemd would then have to decode `\x5C`.
https://github.com/systemd/systemd/blob/bb970e5774580bd8341740edf6ce98b662adc0bf/src/libudev/libudev-util.c#L163-L196

**Comment by JakubVanek** [comment]
That failed (perhaps unsuprisingly) too: `ev3-uart@ev3x5Cx5Cx2dports:in4.service`

**Comment by JakubVanek** [comment]
There is another possibility of fixing it, but I'm not sure of the consequences - if there is a backslash without a quote after it, it would get forwarded as a simple backslash (without "escaping" the next character).
`"\""` -> `"`, `"\x"` -> `\x`

**Comment by TriMoon** [comment]
I'm still puzzled why you are trying to fix something, that is working as intended so far from systemd's side, but is caused by a wrong value passed from an udev rule...
My logic says that the piece of software that passes some value to an other function should provide a correct value for use by that called function of the used external code...

**Comment by TriMoon** [comment]
After checking [L12: ev3-uart@.service](https://github.com/ev3dev/ev3-systemd/blob/ev3dev-stretch/systemd/ev3-uart%40.service#L12), i noticed it uses of `%I` instead of `%i`, see [systemd.unit#Specifiers](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Specifiers)...
Maybe try changing that to see if that helps?

**Comment by JakubVanek** [comment]
> I'm still puzzled why you are trying to fix something, that is [working as intended](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#String%20Escaping%20for%20Inclusion%20in%20Unit%20Names) so far from systemd's side, but is caused by a wrong value passed from an udev rule...
> My logic says that the piece of software that passes some value to an other function should provide a value as expected for use by that called function of the external code...
> 
> Adding extra functionality to the external code is ofcourse helpful to increase usability cases, but not changing current behaviour of it.
> Maybe you can add functionality to accept `SYSTEMD_WANTS_UNESCAPED` for these cases instead ?
> But my recommendation would be to fix the udev rule in such a way that it provides the correct form in the first place...

I think we are referring to two different escaping modes. One is "unquoting/c-unescaping", which converts `"\x"` to `x`. The other one is the one I started to refer as systemd-encoding, which is the `sys-devices` to `sys/devices` one.

**Comment by TriMoon** [comment]
> I think we are referring to two different escaping modes. One is "unquoting/c-unescaping", which converts `"\x"` to `x`. The other one is the one I started to refer as systemd-encoding, which is the `sys-devices` to `sys/devices` one.

Actually `\x` is an escape sequence for hexadecimal...

**Comment by JakubVanek** [comment]
> > I think we are referring to two different escaping modes. One is "unquoting/c-unescaping", which converts `"\x"` to `x`. The other one is the one I started to refer as systemd-encoding, which is the `sys-devices` to `sys/devices` one.
> 
> Actually `\x` is an escape sequence for hexadecimal...

True, that means we have three unquoting modes - there is "systemd-unquoting" = how systemd treats SYSTEMD_WANTS coming from udev = anything preceded by a backslash is forwarded without a backslash and not interpreted as a start/end quote. Then we have C-unescaping that would turn `\x5C` to `\`.

**Comment by TriMoon** [comment]
Did you notice my reply: https://github.com/systemd/systemd/issues/16735#issuecomment-674774203 ?

**Comment by JakubVanek** [comment]
Yes,

> After checking [L12: ev3-uart@.service](https://github.com/ev3dev/ev3-systemd/blob/ev3dev-stretch/systemd/ev3-uart%40.service#L12), i noticed it's use of `%I` instead of `%i`, see [systemd.unit#Specifiers](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Specifiers)...
> Maybe try changing that to see if that helps?

This is the problem. There is a need for both of them: `%I` (the un-encoded path, `ev3-ports:in4`) is needed for the device enablement (attach the line discipline to real device `/dev/tty_ev3-ports:in4`) and `%i` (the original, encoded instance name, `ev3\x2dports:in4`) is needed for the systemd bind path (`sys-subsystem-lego-devices-ev3_uart_host_ev3\x2dports:in4.device` then gets decoded to `sys/subsystem/lego/devices/ev3_uart_host_ev3-ports:in4.device`). 

> I also noticed that you were using `PROGRAM="/bin/systemd-escape $env{LEGO_ADDRESS}"` maybe you should change that to include the `--path` argument to `systemd-escape`?

I tried that and unfortunately it did not have the wanted effect. "Fully correct" (if we didn't need the isolated instance name) variant would be `systemd-escape --path --template=ev3-uart@.service ev3-ports:in4`. That still returns `ev3-uart@ev3\x2dports:in4.service`.

> Which is also mentioned in [systemd.unit#String Escaping for Inclusion in Unit Names](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#String%20Escaping%20for%20Inclusion%20in%20Unit%20Names).
> 
> I have a hunch that (one or both of) these are the real culprit here :wink:

**Comment by JakubVanek** [comment]
Systemd itself (apart from the udev interface) indeed works as expected, if I disable the `BindsTo=` line (otherwise the service gets immediately shut down) and start the `ev3-uart@ev3\x2dports:in4.service`, the driver starts working:
```
$ sudo systemctl start 'ev3-uart@ev3\x2dports:in4.service'
$ sudo systemctl status 'ev3-uart@ev3\x2dports:in4.service'
● ev3-uart@ev3\x2dports:in4.service - LEGO MINDSTORMS EV3 UART Sensors
   Loaded: loaded (/lib/systemd/system/ev3-uart@.service; static; vendor preset: enabled)
   Active: active (running) since Mon 2020-08-17 10:04:17 UTC; 7s ago
  Process: 704 ExecStart=/usr/sbin/ldattach 29 /dev/tty_ev3-ports:in4 (code=exited, status=0/SUCCESS)
 Main PID: 706 (ldattach)
   CGroup: /system.slice/system-ev3\x2duart.slice/ev3-uart@ev3\x2dports:in4.service
           └─706 /usr/sbin/ldattach 29 /dev/tty_ev3-ports:in4

Aug 17 10:04:16 ev3dev systemd[1]: Starting LEGO MINDSTORMS EV3 UART Sensors...
Aug 17 10:04:17 ev3dev systemd[1]: Started LEGO MINDSTORMS EV3 UART Sensors.
```

**Comment by TriMoon** [comment]
Hmm how about transforming the `%I` usage in L12 to use `%i` and transform that into the piece that `ldattach` accepts using an inline-shell script?

**Comment by JakubVanek** [comment]
> But my recommendation would be to fix the udev rule in such a way that it provides the correct form in the first place...

I agree, at this time there are likely setups that depend on the current systemd-unquoting behaviour. I think `OPTIONS+="string_escape=none"` in udev was intended to enable something like the backslash doubling - it *should* affect the `PROGRAM` result in v241 udev (I found that this affects the underscore generation), but I cannot reproduce that. In master, this is no longer the case; the offending `\\` is always transformed to `_\`:
https://github.com/systemd/systemd/blob/bfeb370abcd9a5468ef565a32189d5593a40ce19/src/udev/udev-rules.c#L1669-L1693

**Comment by TriMoon** [comment]
I think its time to change the topic title, so it gets properly noticed by the systemd devs....
We seem to be so far that we agree it's not a problem with `SYSTEMD_WANTS` anylonger, but more a problem with the interaction between udev rules.
Whether it's someting in the udev rule parser inside systemd or not is another topic...

**Comment by JakubVanek** [comment]
> Hmm how about transforming the `%I` usage in L12 to use `%i` and transform that into the piece that `ldattach` expects using an inline-shell script?
> eg.: "sh -c ...."

That would fix the ldattach device path, i.e. `sed s/x2d/-/` without the backslash before `x2d` would do the job. However, the `%i` (expected `ev3\x2dports:in4`, in reality `ev3x2dports:in4`) is put into the systemd `BindsTo=...` path and that is then systemd-de-encoded somewhere inside systemd, yielding a potentially wrong path (that depends into what `SYSTEMD_ALIAS` gets decoded).

However, the approach of wrapping `ldattach` in a helper script is indeed the one that is the way out without patching udev or systemd (the workaround that I was hoping didn't need to be used, as it means not using `LEGO_ADDRESS`). By using `SYSTEMD_WANTS="ev3-uart@.service"` (without `PROGRAM` in udev), systemd internally fills the instance name with sysfs path correctly (instance name like `sys-devices-platform-ev3\x2dports-ev3\x2dports:in4-lego\x2dport-port3-ev3\x2dports:in4:ev3\x2duart\x2dhost`, de-encoded `sys/devices/platform/ev3-ports/ev3-ports:in4/lego-port/port3/ev3-ports:in4:ev3-uart-host`). The encoded `%i` then can be used directly in `BindsTo=` and is guaranteed to be correct. The de-encoded `%I` is then passed to the ldattach wrapper script, where it can be sed'ed to return the `ev3-ports:in4` that we want. I was afraid that this sed script could be quite fragile, but it seems to work OK.

**Comment by TriMoon** [comment]
Is it absolutly required to use `sed` for that transform, because i don't know if the default shell supports replacing strings inside a variable's value a-la-sed like bash has.... fe. `${varname//\x2d/-}`.
But ofcourse using a helper script will open more advanced processing possibilities as what would be possible by just a simple inline script for the unit and still be readable enough :wink:

**Comment by JakubVanek** [comment]
> I think its time to change the topic title, so it gets properly noticed by the systemd devs....
> We seem to be so far that we agree it's not a problem with `SYSTEMD_WANTS` anylonger, but more a problem with the interaction between udev rules.
> Whether it's someting in the udev rule parser inside systemd or not is another topic...

I will open a new issue for it, as it is indeed a conceptually different thing. Yup, I now agree that `SYSTEMD_WANTS` coming from `udev` should want two backslashes, as it does makes sense in the quote context (all unescaped backlashes get eliminated and they prevent the next character from being interpreted as an end quote). I was blind to that at the beginning, thank you for noticing that.

> Is it absolutly required to use `sed` for that transform, because i don't know if the default shell supports replacing strings inside a variable's value a-la-sed like bash has.... fe. `${varname//\x2d/-}`.
> But ofcourse using a helper script will open more advanced processing possibilities as what would be possible by just a simple inline script for the unit and still be readable enough wink

Yes, pattern substitution would certainly work too, but I'm not very familiar with it. For another helper script, `/bin/dash` is used, so that would be the determining factor.

**Comment by JakubVanek** [comment]
Working as expected

**Comment by JakubVanek** [comment]
I have created https://github.com/systemd/systemd/issues/16760 for adding "backslash passthrough mode" to udev's PROGRAM.

**Comment by keszybz** [comment]
> I think the real problem is in the udev rule.
> 
> > ```
> > SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
> >     PROGRAM="/bin/systemd-escape $env{LEGO_ADDRESS}" \
> >     ENV{SYSTEMD_WANTS}="ev3-uart@$result.service", \
> >     ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
> > ```
> 
> Here `PROGRAM` is set to the escaped value of `LEGO_ADDRESS` but the unescaped value is used in setting `SYSTEMD_ALIAS`.

Yeah, I think this is correct.

> But interestingly, when writing the two backslashes directly into `SYSTEMD_WANTS`, it works correctly:
> 
> ```
> SUBSYSTEMS=="lego", ENV{DEVTYPE}=="ev3-uart-host", ACTION=="add", TAG+="systemd", \
>     ENV{SYSTEMD_WANTS}="ev3-uart@ev3\\x2dports:in4.service", \
>     ENV{SYSTEMD_ALIAS}="/sys/subsystem/lego/devices/ev3_uart_host_$env{LEGO_ADDRESS}"
> ```

Yes. A double backslash is needed, because we do unescaping when unquoting, and we need an actual backslash in the unit name.

**Comment by Olf0** [comment]
AFAICS, this is resolved per @JakubVanek's comment https://github.com/ev3dev/ev3dev/issues/1433#issuecomment-674817099.

**Comment by CarloWood** [comment]
Why was this closed? There is no way I can pass an escaped character with `SYSTEMD_WANTS`, and I tried *everything* in the past three hours.

My last attempt looks like this:
```
SUBSYSTEMS=="block", ENV{DEVTYPE}=="partition", ENV{ID_SERIAL_SHORT}=="408D5CBECBBDE7C0E9160666|408D5CBF5F0AE7C0E9150B70|408D5CBECAC3E7C0E9170AEC", PROGRAM="/usr/bin/systemd-escape 'foo-bar'", RUN+="/usr/local/sbin/usbtest '%c'", ENV{SYSTEMD_WANTS}="B@%c.service"
```
This results in `/usr/local/sbin/usbtest` being passed the correct value (`foo\x2dbar`) while the service `B@foox2dbar.service` is being started (removed backslash).

I tried adding backslashes, different quotes, using $result etc etc. Absolutely no way to get the right service started.

Doing `sudo systemctl start 'B@foo\x2dbar.service'` from the command line works as expected, so the bug is clearly with udev and SYSTEMD_WANTS.

Is there a solution for this?!

Using version 257

**Comment by CarloWood** [comment]
If I try `ENV{SYSTEMD_WANTS}="B@foo\\x2dbar.service"`, where I added an extra backslah - then indeed it works.
But if I change the rule to:
```
SUBSYSTEMS=="block", ENV{DEVTYPE}=="partition", ENV{ID_SERIAL_SHORT}=="408D5CBECBBDE7C0E9160666|408D5CBF5F0AE7C0E9150B70|408D5CBECAC3E7C0E9170AEC", PROGRAM="/usr/local/sbin/udev-escape 'foo-bar'", RUN+="/usr/local/sbin/usbtest '%c'", ENV{SYSTEMD_WANTS}="B@%c.service"
```
where `/usr/local/sbin/udev-escape` adds that extra backslash, then the service `B@foo_x2dbar.service` is started. In other words, the double backslash is converted to an underscore, aka it is not working.

Note that `/usr/local/sbin/udev-escape` contains:
```
#! /bin/sh
/usr/bin/systemd-escape "$1" | sed -e 's/\\/\\\\/g'
```
and running from the command line:
```
daniel:/etc/systemd/system>/usr/local/sbin/udev-escape 'foo-bar'
foo\\x2dbar
```

As said before, no number of backslashes makes this work. It is impossible to start a service where %i has a \x2d as escaped character from udev's SYSTEMD_WANTS.

**Comment by dlitz** [comment]
I'm seeing the same behavior as @CarloWood in 256.6.

I'm creating a service that's supposed to be started when a network interface is created, and stopped when it's destroyed:

```
# /etc/systemd/systemd/tc-setup@.service
[Unit]
Description=Traffic Control for interface %I
BindsTo=sys-subsystem-net-devices-%i.device

[Service]
Type=oneshot
ExecStart=/etc/network/tc-setup -i %I
ExecStop=/etc/network/tc-setup -u -i %I
RemainAfterExit=yes
SyslogIdentifier=%N
```

This is my udev rule:

```
SUBSYSTEM=="net", KERNEL!="lo", TAG+="systemd", \
  PROGRAM="/etc/udev/tc-setup-escape --template=tc-setup@.service %k", \
  ENV{SYSTEMD_WANTS}+="%c"
```

This is /etc/udev/tc-setup-escape:

```
#!/bin/sh
set -eu
systemd-escape "$@" | sed -e 's/\\/\\\\/g'
```

Example usage:

```
# /etc/udev/tc-setup-escape --template=tc-setup@.service dummy-one
tc-setup@dummy\\x2done.service
```

When I create a new interface via `ip link add dummy-one type dummy`, I get the following in the journal:

```
Dec 23 16:14:48 petal systemd[1]: Unnecessary job was removed for sys-subsystem-net-devices-dummy_x2done.device - /sys/subsystem/net/devices/dummy_x2done.
Dec 23 16:14:48 petal systemd[1]: Starting tc-setup@dummy_x2done.service - Traffic Control for interface dummy_x2done...
Dec 23 16:14:48 petal tc-setup@dummy_x2done[226249]: tc-setup: no such interface: dummy_x2done
Dec 23 16:14:48 petal tc-setup@dummy_x2done[226249]: tc-setup: No such interface: dummy_x2done
Dec 23 16:14:48 petal systemd[1]: tc-setup@dummy_x2done.service: Main process exited, code=exited, status=1/FAILURE
Dec 23 16:14:48 petal systemd[1]: tc-setup@dummy_x2done.service: Failed with result 'exit-code'.
Dec 23 16:14:48 petal systemd[1]: Failed to start tc-setup@dummy_x2done.service - Traffic Control for interface dummy_x2done.
```

The backslash is listed in udev:

```
# udevadm info /sys/class/net/dummy-one
P: /devices/virtual/net/dummy-one
M: dummy-one
U: net
I: 63
E: DEVPATH=/devices/virtual/net/dummy-one
E: SUBSYSTEM=net
E: INTERFACE=dummy-one
E: IFINDEX=63
E: USEC_INITIALIZED=5354411862061
E: ID_NET_DRIVER=dummy
E: ID_NET_NAMING_SCHEME=v255
E: SYSTEMD_WANTS=tc-setup@dummy_\x2done.service
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: ID_NET_NAME=dummy-one
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/dummy-one
E: TAGS=:systemd:
E: CURRENT_TAGS=:systemd:
```

The failed unit can be seen using systemctl:
```
# systemctl list-units '*dummy*'
  UNIT                                          LOAD   ACTIVE SUB     DESCRIPTION
  sys-devices-virtual-net-dummy\x2done.device   loaded active plugged /sys/devices/virtual/net/dummy-one
  sys-subsystem-net-devices-dummy\x2done.device loaded active plugged /sys/subsystem/net/devices/dummy-one
● tc-setup@dummy_x2done.service                 loaded failed failed  Traffic Control for interface dummy_x2done
```

If I use systemd-escape instead of the script that invokes systemd-escape piped to sed, I get `SYSTEMD_WANTS=tc-setup@dummy\x2done.service` in udevadm, but then systemd looks for a service called `tc-setup@dummyx2done.service`.

The only way to get the double-backslash in SYSTEMD_WANTS seems to be to use `IMPORT{program}` in the udev rule.  (I still contend that is incorrect behavior by systemd---what benefit is there to unescaping the service name listed in SYSTEMD_WANTS in the first place?)

**Comment by JakubVanek** [comment]
I have reopened the issue. However, I will not able to help out with this, as I have forgotten the details and have mostly moved on.

**Comment by CarloWood** [comment]
@JakubVanek Thank you for re-opening. I'm sure I can fix this, I'm just hesitant to put time into third-party projects anymore because my patches are often "not understood" (that is; if it takes me three days to fix a bug, then it requires often two days to understand that my fix is correct and maintainers often don't want to spend more time than 5 minutes; so the patch is then not reviewed, and not applied for 10 years, after which it slowly disappears from reality).

Having stated that - allow me to begin with creating a simple test setup. Hopefully I can get at least one maintainer (you?) to try that and verify that there *is* a problem, before I tackle it.

SIMPLE TEST CASE

Based on the discussion from 2020 I am going to assume that it is desired that
things work iff two backslashes are used in SYSTEMD_WANTS; were the expected
behavior is that every double backslash in SYSTEMD_WANTS ends up as a single
backslash in the %i.

1) Create an executable file `/usr/local/sbin/udev-escaped` with the following content:
```
#! /bin/sh
echo -n 'foo\\x2dbar\\x20\\x21'
```

2) Test it with:
```
~>sudo /usr/local/sbin/udev-escaped | hexdump -C
00000000  66 6f 6f 5c 5c 78 32 64  62 61 72 5c 5c 78 32 30  |foo\\x2dbar\\x20|
00000010  5c 5c 78 32 31                                    |\\x21|
00000015
```
Observe how the output contains two backslashes (5c) (six in total).

3) Create an executable file `/usr/local/sbin/log_service` with the following content:
```
#! /bin/sh
echo "%$1 = \"$2\""
```

4) Test it with:
```
~>sudo /usr/local/sbin/log_service 'i' 'foo\x2dbar\x20\x21'
%i = "foo\x2dbar\x20\x21"
```
Note the use of single quotes and single backslashes. Alternatively run:
```
~>eval /usr/local/sbin/log_service 'i' "$(/usr/local/sbin/udev-escaped)"
%i = "foo\x2dbar\x20\x21"
```
Here the `eval` converted the double backslashes of the output of `udev-escaped` into single backslashes.

5) Create the file `/etc/systemd/system/T@.service` with the following content:
```
[Service]
ExecStart=/bin/systemd-cat -t backslashtest /usr/local/sbin/log_service i '%i'
ExecStop=/bin/systemd-cat -t backslashtest /usr/local/sbin/log_service I '%I'
```

5) Test it as follows:
Run in one terminal `journalctl -b -t backslashtest -f`, and then run in another terminal:
```
$ sudo systemctl start 'T@foo\x2dbar\x20\x21.service'
```
This should return immediately while in the other terminal something is printed like:
```
~>journalctl -b -t backslashtest -f
Dec 24 13:45:06 daniel backslashtest[2398920]: %i = "foo\x2dbar\x20\x21"
Dec 24 13:45:06 daniel backslashtest[2398925]: %I = "foo-bar !"
```
Note how %i is the escaped version and %I is the correct unescaped version.

6) Create the file `/etc/udev/rules.d/42-backslashtest.rules` with something like the following:
```
SUBSYSTEMS=="block", KERNEL=="sd[c-z]1", PROGRAM="/usr/local/sbin/udev-escaped", ENV{SYSTEMD_WANTS}="T@%c.service"
```
Or whatever you think you can trigger by inserting an USB stick (in this case with at least one partition on it),
or otherwise can trigger.

7) Test this as follows:
Run in one terminal `sudo udevadm monitor --property | grep SYSTEMD_WANTS`, and then trigger the above rule
by for example inserting an USB stick with a partition on it. This will result in something like:
```
SYSTEMD_WANTS=T@foo_\x2dbar_\x20_\x21.service
```
Note how this is the same as what @dlitz reported: each backslash is prepended with an underscore.
This is why it can't work.

The expected output is(?):
```
SYSTEMD_WANTS=T@foo\x2dbar\x20\x21.service
```

Also note that the terminal running `journalctl -b -t backslashtest -f` did output:
```
Dec 24 13:58:48 daniel backslashtest[2407094]: %i = "foo_x2dbar_x20_x21"
Dec 24 13:58:48 daniel backslashtest[2407097]: %I = "foo_x2dbar_x20_x21"
```
So, there is another problem and/or the expected output of the above is rather:
```
SYSTEMD_WANTS=T@foo\\x2dbar\\x20\\x21.service
```
I suspect this is what we really want to see.

The bug is then that - when using %c to pass on something with two backslashes,
in fact we pass on `_\` instead of `\\`.

If any maintainer can confirm that they can reproduce and observe the same issue
then I'll propose a patch to fix it.

**Comment by arvidjaar** [comment]
> The following udev rule should start a unit named `ev3-uart@ev3\x2dports:in4.service` if `LEGO_ADDRESS` is `ev3-ports:in4`

Any reason it cannot start a unit named `ev3-uart@ev3-dports:in4.service`?

**Comment by CarloWood** [comment]
@arvidjaar The string passed to a service template should be a systemd-escape -ed  string.  A forward slash '/' is escaped as '-'. If you pass `ev3-uart@ev3-dports:in4.service` then %i inside the service will be `ev3-dports:in4` and when used to decode a devpath that will become `ev3/dports:in4`, screwing up the intended path.

**Comment by CarloWood** [comment]
In the meantime I found a workaround, which is described here:
https://unix.stackexchange.com/a/788649/163895

TLDR: scroll down to "Passing arbitrary information from udev to a service" and then scroll down to "FINAL VERSION" of `69-gold-usb.rules`. The trick is to encode any information NOT using systemd-escape, but using something that produces a string that you CAN pass through a %c with SYSTEMD_WANTS, and using SYMLINK to be able to use that string in unit file:

PROGRAM="/usr/local/sbin/udev-escape '$env{ID_FS_UUID} $env{ID_FS_TYPE}'", \
SYMLINK+="usb/%c", \
ENV{SYSTEMD_WANTS}="mount-gold@usb-%c.service"

Now you can use in `mount-gold@.service`: `BindsTo=dev-%i.device` etc, while, after passing `%i` to a script you can get back whatever you encoded in it, for example:

```
# Strip the leading "usb-" from the argument.
ENCODED_DATA=${1:4}
# Decode the data that was passed.
DATA="$(/usr/local/sbin/udev-unescape $ENCODED_DATA)"
# Extract the value of ID_FS_UUID (the first argument).
ID_FS_UUID=${DATA%%\ *}
# Extract the value of ID_FS_TYPE (the second argument).
ID_FS_TYPE=${DATA#*\ }
```

**Comment by TriMoon** [comment]
> [!IMPORTANT]
@CarloWood 
> >The bug is then that - when using %c to pass on something with two backslashes,
> >in fact we pass on `_\` instead of `\\`.

So the topic title should actually read:
- **"BUG udev: `%c` does not preserve backslashes properly"**

Because when that is fixed, it does not matter if you use the value of `%c` in `SYSTEMD_WANTS` or something else.
The problem is not in the interpretation of the result, the problem is actually the generation of the result for `%c` that seems to be bugged.

This topic is very old, and there has been a lot of changes since the 241 version it was reported for.
So it would be good to have info on current state version behavior also.
_(I don't make use of this behavior myself, else i would provide that info myself)_

**Comment by arvidjaar** [comment]
> `/usr/local/sbin/udev-escaped` with the following content:
> 
> ```
> #! /bin/sh
> echo -n 'foo\\x2dbar\\x20\\x21'
> ```
...
> `PROGRAM="/usr/local/sbin/udev-escaped", ENV{SYSTEMD_WANTS}="T@%c.service"`

`IMPORT` should preserve backslashes, so changing it into
```
#!/bin/sh
echo -n 'SYSTEMD_WANTS=T@foo\\x2dbar\\x20\\x21.service'
```
and replacing the rule with
```
IMPORT{program}="/usr/local/sbin/udev-escaped"
```
works without jumping through the hoops.

**Comment by CarloWood** [comment]
@arvidjaar Merry Xmas. Using `%c{1+}` makes no difference. Using IMPORT works!

I am glad about this, because I just found out that there is a problem with using the SYMLINK:
when running cryptsetup to create a /dev/mapper/foo of that, the SYMLINK .device becomes inactive / dead?!
Must be a bug... it makes no sense (all other block device .device's just stay, it is JUST the symlink device).
Hence, using `BindsTo=dev-%i.device` where %i us a SYMLINK doesn't do what I want: that terminates the
service in the middle of running cryptsetup.

That being said, I'm still clueless on how to properly decrypt a partition where the key is gotten through gpg-agent anyway: the official way seems to be to add something to /etc/crypttab and then use systemd-cryptsetup@.service?
That requires a path in /etc/crypttab from which the passphrase can be read however.  Maybe I can make it a socket, and write a program that gets the passphrase from my YubiKey (which involves popping up windows that tell me to insert the key and type in a PIN) when something is trying to read from it... That socket/path itself then can be created as a result of running a "daemon" thus, which could be started by systemd too. All in all, this is going to be weeks and weeks of work :(

**Comment by CarloWood** [comment]
@TriMoon 

> So the topic title should actually read:
> 
> * **"BUG udev: `%c` does not preserve backslashes properly"**
> 
> Because when that is fixed, it does not matter if you use the value of `%c` in `SYSTEMD_WANTS` or something else. The problem is not in the interpretation of the result, the problem is actually the generation of the result for `%c` that seems to be bugged.

That is correct. Maybe I should have create a new issue, but I didn't know it was this a few days ago.
At any rate, SYSTEMD_WANTS seems to convert '\\' into '\', which is apparently what we want.

What remains is that '%c' converts '\\' into '_\', or at least SYSTEMD_WANTS doesn't have to do anything with that.
This is the current state of affairs.

> @CarloWood What do you get when you use fe. `ENV{SYSTEMD_WANTS}="T@%c{1+}.service"` or similar to prevent splitting the result into different parts with spaces between them 🤔 _(My guess is that a `\` somehow "triggers" the generator of `%c` to interpret it as a different part by default)_

Sorry I didn't investigate that in-depth because I don't understand what you are referring to. I merely tried exactly that `ENV{SYSTEMD_WANTS}="T@%c{1+}.service"` and the result is the same as when using ``ENV{SYSTEMD_WANTS}="T@%c.service"``.

**Comment by arvidjaar** [comment]
> '%c' converts '\\' into '_\'

There is no way to change it now without risk of breaking unknown number of external rules. You will need to introduce new substitution (or modifier to the existing one) with different semantic. But then you can simply use `IMPORT` which works just fine and does exactly what you need.


> SYSTEMD_WANTS seems to convert '\\' into '\'

One could argue that it should not do it, and it was a mistake. But same consideration - it is no more possible to change without breaking compatibility. It would need new variable, something like `SYSTEMD_WANTS_UNESAPED` with different semantic.

What could be done, is documenting what the `systemd` does with the value of `SYSTEMD_WANTS`. The documentation only says
> should be a space-separated list of valid unit names

which makes impression that `foo\x2dbar.service` works - while in reality it does not. **That** is a real bug IMHO. `foo\x2dbar.service` **is** a "valid unit name" but does not work.

BTW, `%c` would also preserve such value ... so avoiding the whole issue.

**Comment by JakubVanek** [comment]
> One could argue that it should not do it, and it was a mistake. But same consideration - it is no more possible to change without breaking compatibility. It would need new variable, something like SYSTEMD_WANTS_UNESAPED with different semantic.

Interestingly, I think that it used to work that way. If I remember things correctly, I originally created this issue because ev3dev's [systemd integration](https://github.com/ev3dev/ev3-systemd/) broke when upgrading from Debian Stretch (systemd v232) to Debian Buster (systemd v241). IOW: I think that in systemd <=v232, `SYSTEMD_WANTS` integration maybe worked in ev3dev even without the commit https://github.com/ev3dev/ev3-systemd/commit/a672151f8c8b8cd9cca2a635bb4b197cbaa74afe (which is another possible workaround). It is my fault that I have not formulated this into the original issue submission.

**Comment by JakubVanek** [comment]
I have found another possible workaround via a mention in the original ev3dev issue: https://github.com/Olf0/crypto-sdcard/commit/0d673785217e87e81bd3a44c9108edfbf65929de .

**Comment by JakubVanek** [comment]
Confirmed: this is an old regression.

How to reproduce the original (non-regressed) behaviour:

 1. Install Debian 9.13 into a `virt-manager` VM. The systemd package version in this distro is `232-25+deb9u12`.
 2. Log into the VM.
 3. Create a file `/etc/systemd/system/printer@.service` with the following contents:
     ```
     [Service]
     ExecStart=/bin/sh -c "/bin/echo 'i=%i' 'I=%I' | /usr/bin/systemd-cat -t systemd-repro"
     ```
 4. Create a file `/etc/udev/rules.d/42-backslashtest.rules` with the following contents:
     ```
     SUBSYSTEMS=="block", KERNEL=="sd[a-z]1", PROGRAM="/bin/systemd-escape triggered-by-udev", ENV{SYSTEMD_WANTS}="printer@%c.service"
     ```
 5. Try to run the service directly:
     ```
     $ sudo systemctl start 'printer@test\x2drun.service'
     ```
 6. Plug in a USB drive (it is sufficient to use SPICE USB forwarding in virt-manager). This should create a new /dev/sdX1 device node and trigger the udev rule.
 7. Observe that the backslash is passed correctly and (un)escaping works as desired:
    ```
    $ sudo journalctl -t systemd-repro
    -- Logs begin at Thu 2024-12-26 18:57:39 CET, end at Thu 2024-12-26 18:58:20 CET. --
    pro 26 18:58:08 debian9-test-vm systemd-repro[468]: i=test\x2drun I=test-run
    pro 26 18:58:20 debian9-test-vm systemd-repro[493]: i=triggered\x2dby\x2dudev I=triggered-by-udev
    ```

How to reproduce the regressed behaviour:
 1. Install a Ubuntu 24.04 into a virt-manager VM. In this distro, the systemd version is `255.4-1ubuntu8.4`.
 2. Follow the rest of the reproduction steps for Debian 9.
 3. Observe that the output of the `sudo journalctl -t systemd-repro` command is different:
     ```
     pro 26 19:00:47 nlcm-ros-vm systemd-repro[1155]: i=test\x2drun I=test-run
     pro 26 19:01:03 nlcm-ros-vm systemd-repro[1179]: i=triggeredx2dbyx2dudev I=triggeredx2dbyx2dudev
     ```

**Comment by JakubVanek** [comment]
I don't think the regression is in udev after all (albeit the ultimate fix might need to be in udev, to avoid breaking more stuff). The variables that udev passes to systemd are the same between the old and new systemd versions:

On Debian 9.13, the following is printed when the USB drive is forwarded to the VM:
```
$ sudo udevadm monitor --property | grep SYSTEMD_WANTS
SYSTEMD_WANTS=printer@triggered\x2dby\x2dudev.service
```

On Ubuntu 24.04, the following is printed when the USB drive is forwarded to the VM:
```
$ sudo udevadm monitor --property | grep SYSTEMD_WANTS
SYSTEMD_WANTS=printer@triggered\x2dby\x2dudev.service
```

**Comment by JakubVanek** [comment]
My guess is that the regression is present in systemd since https://github.com/systemd/systemd/commit/ceed8f0c8b9a46300eccd1afa2dd8d3c2cb6b47c - the `FOREACH_WORD_QUOTED()` and `extract_first_word()` mechanisms handle backslashes differently.

**Comment by arvidjaar** [comment]
> systemd <=v232

> My guess is that the regression is present in systemd since [ceed8f0](https://github.com/systemd/systemd/commit/ceed8f0c8b9a46300eccd1afa2dd8d3c2cb6b47c) - the `FOREACH_WORD_QUOTED()` and `extract_first_word()` mechanisms handle backslashes differently.

Yes. The `FOREACH_WORD_QUOTED` behaved rather interesting. The `\` quoted the next character but was left in the result verbatim. So, `f\oo\ b\ar` resulted in the actual string `f\oo\ b\ar`, while `f\oo b\ar` gave `f\oo` and `b\ar`. `extract_first_word()` behaves conventionally - quote character itself is removed.

**Comment by JakubVanek** [comment]
> > systemd <=v232
> 
> > My guess is that the regression is present in systemd since [ceed8f0](https://github.com/systemd/systemd/commit/ceed8f0c8b9a46300eccd1afa2dd8d3c2cb6b47c) - the `FOREACH_WORD_QUOTED()` and `extract_first_word()` mechanisms handle backslashes differently.
> 
> Yes. The `FOREACH_WORD_QUOTED` behaved rather interesting. The `\` quoted the next character but was left in the result verbatim. So, `f\oo\ b\ar` resulted in the actual string `f\oo\ b\ar`, while `f\oo b\ar` gave `f\oo` and `b\ar`. `extract_first_word()` behaves conventionally - quote character itself is removed.

Thank you, this reassured me a bit.

I have just checked that applying https://github.com/systemd/systemd/pull/16737 onto systemd `255.4-1ubuntu8.4` restores the old escaping behaviour (at least for the regression reproducer).

However, I am not entirely sure that merging the PR is the right decision now (given that the new behaviour is present for ~8 years now). It would make `systemd-escape` easier to use inside udev rules (no need to use the `IMPORT{program}` workaround), but it might indeed break some newer udev rules.

**Comment by Olf0** [comment]
IMO, `systemd` is not behaving as specified, hence this is a bug rsp. a regression which should be fixed.

Furthermore, the current state is outright broken, one cannot propagate arbitrary strings via `systemd-escape` as promised.

More references / sibling issues:
- https://github.com/systemd/systemd/issues/7109#issuecomment-338959276
- https://github.com/systemd/systemd/issues/7587
- https://github.com/ev3dev/ev3dev/issues/1433#issuecomment-674817099
- Loosely related: https://github.com/systemd/systemd/issues/10321

**Comment by TriMoon** [comment]
I agree that this should be fixed.

No matter how old a bug is, or how long it's already in circulation, IMHO it is unacceptable to "leave it as-is".

Some big commercial companies, which we won't mention here, might get away with bad code for years because the public has no way of slapping them in the head...
But community developed software, like this, should fix any and every bug that is found when possible asap...

The strength of community software is also it's weakness, one coder can introduce a bug or regression that another coder (or same) can fix later.
But leaving bugs as-is, is just asking for a code-base that will kill itself in the long run due to those bugs...

Unless we want to become a joke in coding like those companies...

**Comment by JakubVanek** [comment]
> I mean - if you hardcode two backslahes into SYSTEMD_WANTS it works.

Actually, if https://github.com/systemd/systemd/pull/35768 or https://github.com/systemd/systemd/pull/16737 is merged, it will stop working the way it works with current systemd.

* Current behaviour:
  * `SYSTEMD_WANTS=printer@triggered\x2dby\x2dudev.service` (passed via udev `PROGRAM`) leads to `%i` being `triggeredx2dbyx2dudev` and `%I` being `triggeredx2dbyx2dudev`.
  * `SYSTEMD_WANTS=printer@triggered\\x2dby\\x2dudev.service` (passed via udev `IMPORT{program}`) leads to `%i` being `triggered\x2dby\x2dudev` and `%I` being `triggered-by-udev`.
  * The way to call `systemd-escape` for SYSTEMD_WANTS in udev rules seems to be via the `IMPORT{program}="<wrapper env-generating script>"` workaround - this doubles the backslashes.
* Behaviour of systemd <= v232 and of the pull requests:
  * `SYSTEMD_WANTS=printer@triggered\x2dby\x2dudev.service` (passed via udev `PROGRAM`) leads to `%i` being `triggered\x2dby\x2dudev` and `%I` being `triggered-by-udev`.
  * `SYSTEMD_WANTS=printer@triggered\\x2dby\\x2dudev.service` (passed via udev `IMPORT{program}`) leads to `%i` being `triggered\\x2dby\\x2dudev`. If `%I` is present in the systemd unit, systemd will not start it at all and a message "Failed to resolve unit specifiers: Invalid argument" will be shown.
  * The way to call `systemd-escape` for SYSTEMD_WANTS in udev rules is to run it via the `PROGRAM="/usr/bin/systemd-escape something"` directive - no backslash doubling is needed.


There is another technicality regarding the `%c` formatting specifier: I think that the problem is not with it (i.e. the code handling `%c` does not replace `\\` with `_\`). The problem is in the `PROGRAM` directive handler [here](https://github.com/systemd/systemd/blob/842a674d4f7fb69af28e67f447c1250383e50f41/src/udev/udev-rules.c#L2157). To fix it, there would have to be a new configuration directive similar to `string_escape` (see https://man7.org/linux/man-pages/man7/udev.7.html) that would disable this post-processing; see https://github.com/systemd/systemd/issues/16760.

I think that https://github.com/systemd/systemd/issues/16760 is not blocked by anything; if you want, I can reopen it as well.

**Comment by JakubVanek** [comment]
Relevant context from a sibling issue:

*@CarloWood in https://github.com/systemd/systemd/issues/16760#issuecomment-2564343067:*
> PS I just tried hard to find anything with Google that refers to / documents the use of SYSTEMD_WANTS, and the only things I can find are either posts from more than 9 years ago, people not using `@`, or people using it to pass `%k`, `%p`, `$name` or `$devnode`. I can't find ANYONE talking about having to escape backslashes or doing anything special. So, I'm inclined to think that fixing the bug won't have a large impact on current configurations.

*@JakubVanek in https://github.com/systemd/systemd/issues/16760#issuecomment-2564521840:*
> This sounds good. I now also tried to do a quick GitHub code search (https://github.com/search?q=PROGRAM+SYSTEMD_WANTS+%40+path%3A*.rules&type=code) and it seems that the remaining people are still relying on PROGRAM + systemd-escape the way it worked in systemd <=v232.

**Comment by TriMoon** [comment]
Hope your fix works @yuwata 👍
